### DISCUSSION TITLE
Provide own instance of IPhase4PeppolIncomingSBDHandlerSPI

### DISCUSSION THREAD

**Discussion by Florianisme** [phax/phase4#115] [Ideas]
Hi Philip,

I am currently using your library in a Spring Boot Project. I have provided an implementation of the IPhase4PeppolIncomingSBDHandlerSPI Interface through SPI which is working fine.

It was however always a headache for me to do it this way because I have to use Autoinjected Dependencies in that class. That meant that I had to keep a reference to the Spring Context and manually inject my dependencies that way.

It would be way easier if I could register a Provider or Factory instance in your library which then provides an instance of my implementation for your library to use. This class can be kept completely Spring-free.

What do you think about this idea?

**Comment by phax** [comment]
Hi Florian, Thanks for reaching out to me.
I took the liberty and converted this to a discussion rather then an issue. If we agree on the solution, we may end up in an issue :)

I think I understand your considerations, even though I am not a Spring expert (please bare that in mind when you read the rest of my suggestion).

Making the change you suggested would mean a breaking change for all existing users. Instead I would like to suggest a different solution that just takes one more level of indirection.

* So the proposal is to stay with the SPI solution
* You keep your SPI handler implementation class (calling it `FlorianSPI` :) )
* Inside `FlorianSPI` you add a static method "registerHandler (yourInterface)` - the interface should be a subset of SPI interface (you can even reduce the parameters to the ones you need)
* Everytime the SPI part of `FlorianSPI` is invoked, you just forward the call to your registered handler from the previous step

With this solution, you are losing a few nanonseconds for the indirection, but finally you have an object that is fully injected and you can do whatever you need.

How does that sound to you?

**Reply by Florianisme** [reply]
Hi Philip,

thanks for your time!
That is actually a really good Idea. I have implemented it and it works like a charm. 

If there was a general interest in doing something similar in other implementations of your library we could have thought about integrating this mechanism in the library itself. 
But for now that's fine for me, especially as it might require a breaking change in your code or at the very least behave like a well-documented fallback or something :)

Thanks anyways, you may close the discussion!

**Reply by phax** [reply]
Hi Florian,
Thanks for the positive response. Happy to hear that it works :)

I will try to include something like this as an example code into the "spring-boot-demo" - it may makes sense to point others to this solution.

This discussion cannot be closed - it stays open. If would be nice if you could select my answer as the "selected answer" so that others reading the thread are directly routed to the selected solution.

Happy coding :)

**Reply by phax** [reply]
Update: see commit https://github.com/phax/phase4/commit/b40ef99e0045ed51a094acecdbd960820dd819d6 for the respecitve commit

**Reply by Florianisme** [reply]
I actually can't select your response as the "selected answer" - at least I don't see any button for it anywhere :D

**Reply by phax** [reply]
Honestly I don't know either. I just saw it on https://github.com/phax/phase4/discussions/113

**Reply by Florianisme** [reply]
Yeah I've seen that. 

Maybe it is because you opened the discussion from my issue. Even though it shows my name as the author. Strange.
